front/src/modules/trainschedule/components/DriverTrainSchedule/DriverTrainScheduleExportCSV.ts

```diff
@@ -0,0 +1,148 @@
+    speedLimitList.map((d: SpeedPosition) => d.position),
+    position
+  );
+  return speedLimitList[bisector].speed || 0;
+};
+
+// Add OPs inside speedsteps array, gather speedlimit with stop position, and sort the array along position before return
+const overloadWithOPs = (train: Train, baseOrEco: BaseOrEcoType): PositionSpeedTimeOP[] => {
+  const trainRegime = train[baseOrEco];
+  if (trainRegime) {
+    const trainsRegimeWithOPs = trainRegime.speeds.concat(
+      trainRegime.stops.map((stop) => ({
+        position: stop.position,
+        speed: interpolateValue(stop.position, trainRegime.speeds, 'speed'),
+        time: stop.time,
+        op: stop.name,
+        ch: stop.ch,
+        lineCode: stop.line_code,
+        trackName: stop.track_name,
+      }))
+    );
+    return trainsRegimeWithOPs.sort((stepA, stepB) => stepA.position - stepB.position);
+  }
+  return [];
+};
+
+const overloadWithSpeedLimits = (
+  trainRegime: PositionSpeedTimeOP[],
+  speedLimits: SpeedPosition[]
+): PositionSpeedTimeOP[] | undefined => {
+  const trainsRegimeWithOPs = trainRegime.concat(
+    speedLimits.map((step) => ({
+      position: step.position,
+      speed: interpolateValue(step.position, trainRegime, 'speed'),
+      speedLimit: step.speed,
+      time: interpolateValue(step.position, trainRegime, 'time'),
+    }))
+  );
+  return trainsRegimeWithOPs.sort((stepA, stepB) => stepA.position - stepB.position);
+};
+
+function spreadTrackAndLineNames(steps: CSVData[]): CSVData[] {
+  let oldTrackName = '';
+  let oldLineCode = '';
+  const newSteps: CSVData[] = [];
+  steps.forEach((step) => {
+    const newTrackName =
+      oldTrackName !== '' && step.trackName === '' ? oldTrackName : step.trackName;
+    const newLineCode = oldLineCode !== '' && step.lineCode === '' ? oldLineCode : step.lineCode;
+    newSteps.push({
+      ...step,
+      trackName: newTrackName,
+      lineCode: newLineCode,
+    });
+    oldTrackName = newTrackName;
+    oldLineCode = newLineCode;
+  });
+  return newSteps;
+}
+
+function createFakeLinkWithData(train: Train, baseOrEco: BaseOrEcoType, csvData: CSVData[]) {
+  const currentDate = new Date();
+  const header = `Date: ${currentDate.toLocaleString()}\nName: ${train.name}\nType:${baseOrEco}\n`;
+  const keyLine = `${Object.values(CSVKeys).join(';')}\n`;
+  const csvContent = `data:text/csv;charset=utf-8,${header}\n${keyLine}${csvData.map((obj) => Object.values(obj).join(';')).join('\n')}`;
+  const encodedUri = encodeURI(csvContent);
+  const fakeLink = document.createElement('a');
+  fakeLink.setAttribute('href', encodedUri);
+  fakeLink.setAttribute('download', `export-${train.name}-${baseOrEco}.csv`);
+  document.body.appendChild(fakeLink);
+  fakeLink.click();
+  document.body.removeChild(fakeLink);
+}
+
+export default function driverTrainScheduleExportCSV(train: Train, baseOrEco: BaseOrEcoType) {
+  const trainRegime = overloadWithSpeedLimits(overloadWithOPs(train, baseOrEco), train.vmax);
```

Comment:
Regime is quite confusing because it's a more complex object in the code, it's not only speeds. I think we should rename the const trainSpeeds :)
```suggestion
  const trainSpeeds = overloadWithSpeedLimits(overloadWithOPs(train, baseOrEco), train.vmax);
```

Reply:
Exactly: the object obtained is not only speeds, but stops, speedlimit and time. I think it's perfectly what it is supposed to be ;)